File: src/audioWfc/wfc/constraints/PlagalCadenceSoftConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.wfc.constraints.concepts.GenericCadenceSoftConstraint;
import audioWfc.wfc.grabbers.ChordQualityConstantGrabber;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.wfc.grabbers.IntegerConstantGrabber;
import audioWfc.musicTheory.Key;
import audioWfc.musicTheory.chords.ChordQuality;

public class PlagalCadenceSoftConstraint extends GenericCadenceSoftConstraint {

    public PlagalCadenceSoftConstraint(double factor, Grabber<Key> keyGrabber){
        super(factor, keyGrabber, new IntegerConstantGrabber(5), new ChordQualityConstantGrabber(ChordQuality.MAJOR),
                new IntegerConstantGrabber(0), new ChordQualityConstantGrabber(ChordQuality.MAJOR));
    }

    @Override
    public String name() {
        return "Plagal cadences";
    }

    @Override
    public String configText() {
        return keyGrabber.configText() + ", Bonus: " + factor;
    }
}

File: vite-project/src/wfc/constraints/cadences/PlagalCadenceSoftConstraint.ts
import { ConcreteChordQuality } from "../../../music_theory/Chord"
import { MusicalKey } from "../../../music_theory/MusicalKey"
import { Grabber } from "../../Grabber"
import { constantGrabber } from "../../grabbers/constantGrabbers"
import { chordConstraintTypeToName } from "../constraintUtils"
import { GenericCadenceSoftConstraint } from "./GenericCadenceSoftConstraint"

export class PlagalCadenceSoftConstraint extends GenericCadenceSoftConstraint {
	name = chordConstraintTypeToName.get("PlagalCadenceSoftConstraint") as string
	configText = () => `Key: ${this.keyGrabber.configText()}, Bonus: ${this.bonus}`
	constructor(bonus: number, keyGrabber: Grabber<MusicalKey>) {
		super(bonus, constantGrabber(5), constantGrabber(0), constantGrabber("major" as ConcreteChordQuality), constantGrabber("major" as ConcreteChordQuality), keyGrabber)
	}
}